src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetVolumePathName.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+
+internal static partial class Interop
+{
+    internal static partial class Kernel32
+    {
+        [LibraryImport(Libraries.Kernel32, EntryPoint = "GetVolumePathNameW", SetLastError = true, StringMarshalling = StringMarshalling.Utf16)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        private static unsafe partial bool _GetVolumePathName(char* lpszFileName, char* lpszVolumePathName, int cchBufferLength);
+
+        public static unsafe int GetVolumePathName(string fileName, out string? volumePath)
+        {
+            fileName = PathInternal.EnsureExtendedPrefixIfNeeded(fileName); //todo: unsure if this is needed for this API
+            fixed (char* lpszFileName = fileName)
+            {
+                // Allocate output buffer on the stack initially.
+                const int stackAllocation = 512;
+                Span<char> volumePathBuffer = stackalloc char[stackAllocation + 1]; // +1 to ensure a \0 at the end, todo: is this necessary
+                int bufferSize = stackAllocation;
+
+                // Loop until the buffer's big enough.
+                while (true)
+                {
+                    fixed (char* lpszVolumePathName = volumePathBuffer)
+                    {
+                        // Try to get the volume name, will succeed if the buffer's big enough.
+                        if (_GetVolumePathName(lpszFileName, lpszVolumePathName, volumePathBuffer.Length - 1))
+                        {
+                            volumePath = new string(lpszVolumePathName);
+                            return 0;
+                        }
+
+                        // Check if the error was that the buffer is not large enough.
+                        int error = Marshal.GetLastWin32Error();
+                        if (error == Interop.Errors.ERROR_INSUFFICIENT_BUFFER) //todo: check this is the correct error
+                        {
+                            // Create a new buffer and try again.
+                            // todo: use array pool and check for overflow
+                            volumePathBuffer = new char[bufferSize *= 2];
```

Comment:
Todo: Should probably use `ArrayPool` with the max path length.

Reply:
We often use the internal ValueStringBuilder to handle the case where a stackalloc needs to be swapped for a pooled array. @jkoritzinsky @stephentoub any reason it's not used in interop? I don't see use in Common/src/Interop/**.

Reply:
instead I see us do it manually eg
https://github.com/danmoseley/runtime/blob/0cf1f09d6a90d873e206a2357feafe5f05e56b43/src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs#L37-L69

Reply:
I can't think of any reason. It looks like it's generally not used in the Interop folder, but it's used in code that calls into the Interop class where the method in the Interop class is just the P/Invoke (no additional logic) like `Interop.Kernel32.GetCurrentDirectory`.

Reply:
Great here's an example @hamarb123 
https://github.com/dotnet/runtime/blob/0cf1f09d6a90d873e206a2357feafe5f05e56b43/src/libraries/System.Private.CoreLib/src/System/Environment.Windows.cs#L20-L41
and you pull it into your project like this if needed
https://github.com/dotnet/runtime/blob/0cf1f09d6a90d873e206a2357feafe5f05e56b43/src/libraries/System.Console/src/System.Console.csproj#L29

Incidentally, that example appears to not be consistently disposing the VSB, thus leaking any array from the pool?